src/d3d9/shaders/d3d9_fixed_function_frag.frag

```diff
@@ -0,0 +1,667 @@
+}
+
+bool resultIsTemp(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[1], 18, 1) != 0;
+}
+bool globalSpecularEnabled() {
+    return bitfieldExtract(data.Stages[0].Primitive[1], 19, 1) != 0;
+}
+
+
+vec4 calculateFog(vec4 vPos, vec4 oColor) {
+    vec3 fogColor = vec3(rs.fogColor[0], rs.fogColor[1], rs.fogColor[2]);
+    float fogScale = rs.fogScale;
+    float fogEnd = rs.fogEnd;
+    float fogDensity = rs.fogDensity;
+    D3DFOGMODE fogMode = specUint(SpecPixelFogMode);
+    bool fogEnabled = specBool(SpecFogEnabled);
+    if (!fogEnabled) {
+        return oColor;
+    }
+
+    float w = vPos.w;
+    float z = vPos.z;
+    float depth = z * (1.0 / w);
+    float fogFactor;
+    switch (fogMode) {
+        case D3DFOG_NONE:
+            fogFactor = in_Fog;
+            break;
+
+        // (end - d) / (end - start)
+        case D3DFOG_LINEAR:
+            fogFactor = fogEnd - depth;
+            fogFactor = fogFactor * fogScale;
+            fogFactor = spvNClamp(fogFactor, 0.0, 1.0);
+            break;
+
+        // 1 / (e^[d * density])^2
+        case D3DFOG_EXP2:
+        // 1 / (e^[d * density])
+        case D3DFOG_EXP:
+            fogFactor = depth * fogDensity;
+
+            if (fogMode == D3DFOG_EXP2)
+                fogFactor *= fogFactor;
+
+            // Provides the rcp.
+            fogFactor = -fogFactor;
+            fogFactor = exp(fogFactor);
+            break;
+    }
+
+    vec4 color = oColor;
+    vec3 color3 = color.rgb;
+    vec3 fogFact3 = vec3(fogFactor);
+    vec3 lerpedFrog = mix(fogColor, color3, fogFact3);
+    return vec4(lerpedFrog.r, lerpedFrog.g, lerpedFrog.b, color.a);
+}
+
+
+// [D3D8] Scale Dref to [0..(2^N - 1)] for D24S8 and D16 if Dref scaling is enabled
+vec4 adjustDref(vec4 texCoord, uint referenceComponentIndex, uint samplerIndex) {
+    float reference = texCoord[referenceComponentIndex];
+    uint drefScaleFactor = specUint(SpecDrefScaling);
+    if (drefScaleFactor != 0) {
+        float maxDref = 1.0 / (float(1 << drefScaleFactor) - 1.0);
+        reference *= maxDref;
+        texCoord[referenceComponentIndex] = reference;
+    }
+    if (specBool(SpecSamplerDrefClamp, samplerIndex)) {
+        texCoord[referenceComponentIndex] = clamp(reference, 0.0, 1.0);
+    }
+    return texCoord;
+}
+
+
+vec4 calculateBumpmapCoords(uint stage, vec4 baseCoords, vec4 previousStageTextureVal) {
+    uint previousStage = stage - 1;
+
+    vec4 coords = baseCoords;
+    [[unroll]]
+    for (uint i = 0; i < 2; i++) {
+        float tc_m_n = coords[i];
+        vec2 bm = vec2(sharedData.Stages[previousStage].BumpEnvMat[i][0], sharedData.Stages[previousStage].BumpEnvMat[i][1]);
+        vec2 t = previousStageTextureVal.xy;
+        float result = tc_m_n + dot(bm, t);
+        coords[i] = result;
+    }
+    return coords;
+}
+
+
+uint loadSamplerHeapIndex(uint samplerBindingIndex) {
+    uint packedSamplerIndex = packedSamplerIndices[samplerBindingIndex / 2u];
```

Comment:
i swear one day amd's windows driver will support 16-bit push constants so we can get rid of this stupid packing